### PR TITLE
 Enable pylint plugin pylint.extensions.docparams and resolve errors

### LINKED ISSUE #1192: Enable pylint plugin pylint.extensions.docparams
from its description

    Checker for Sphinx, Google, or Numpy style docstrings

    * Check that all function, method and constructor parameters are mentioned
      in the params and types part of the docstring.  Constructor parameters
      can be documented in either the class docstring or ``__init__`` docstring,
      but not both.
    * Check that there are no naming inconsistencies between the signature and
      the documentation, i.e. also report documented parameters that are missing
      in the signature. This is important to find cases where parameters are
      renamed only in the code, not in the documentation.
    * Check that all explicitly raised exceptions in a function are documented
      in the function docstring. Caught exceptions are ignored.

    Activate this checker by adding the line::

        load-plugins=pylint.extensions.docparams

**Definition of Done**
- docparams extension is enabled in `make pylint`
- All errors reported from it are fixed (around 15 at the time of writing).
- Pylint error labels to look for: multiple-constructor-doc, missing-raises-doc, redundant-returns-doc, redundant-yields-doc, missing-return-doc, missing-return-type-doc, missing-yield-type-doc, missing-param-doc, missing-type-doc, differing-param-doc, differing-type-doc, 
- pull request opened against this repository
- pull request is closed

**More info**
- https://kiwitcms.readthedocs.io/en/latest/contribution.html

*This issue is part of Kiwi TCMS open source bounty program. For more information see the link(s) in [bounty-program milestone](https://github.com/kiwitcms/Kiwi/milestone/4)*

### PR DESCRIPTION
**Pull request by brymut**
Resolves #1192. Enable pylint plugin pylint.extensions.docparams and resolve errors

Can I get some clarification on a few things I'm not clear on to complete the doc-params : 

- I'm not quite sure what `op_type` is in:
https://github.com/kiwitcms/Kiwi/blob/dd295479eae3b73bb1b99811c610f5d20657f5ca/tcms/rpc/utils.py#L90
and why its used here and why the TypeError is raised.

- I'm not familiar with how to document kwargs in docparams. May I please get a sample to go off on, for example in:
https://github.com/kiwitcms/Kiwi/blob/dd295479eae3b73bb1b99811c610f5d20657f5ca/tcms/rpc/api/auth.py#L14

### CODE REVIEW

**Comment by atodorov** [comment]
> Resolves #1192. Enable pylint plugin pylint.extensions.docparams and resolve errors
> 
> Can I get some clarification on a few things I'm not clear on to complete the doc-params :
> 
>     * I'm not quite sure what `op_type` is in:
>       https://github.com/kiwitcms/Kiwi/blob/dd295479eae3b73bb1b99811c610f5d20657f5ca/tcms/rpc/utils.py#L90
>       

IDK what that actually is. It is legacy code used in object serialization which will be removed at some point but that affects a rather big portion of the app. I can see this whole module uses different style documentation strings. One option is to use kind of dummy descriptions just to satisfy the linter and the other one is to disable some of the checks for this module. None is ideal but the alternative is debugging and figuring out what every parameter does which I don't see the value in.

>       and why its used here and why the TypeError is raised.
> 
>     * I'm not familiar with how to document kwargs in docparams. May I please get a sample to go off on, for example in:
>       https://github.com/kiwitcms/Kiwi/blob/dd295479eae3b73bb1b99811c610f5d20657f5ca/tcms/rpc/api/auth.py#L14

`:param kwargs: description` ???

**Comment by brymut** [comment]
Not quite sure how _kwargs is being used here:
https://github.com/kiwitcms/Kiwi/blob/dd295479eae3b73bb1b99811c610f5d20657f5ca/tcms/testplans/models.py#L107

**Comment by atodorov** [comment]
> Not quite sure how _kwargs is being used here:

Underscore as the name prefix indicates the variable isn't used. IIRC this method received form fields/HTTP request params so this is a catch-all container for any extra input which may be present but we don't care about. If you need to document it this is the explanation for it.

File: tcms/issuetracker/base.py [hunk 1]
```diff
@@ -15,9 +15,13 @@ class IntegrationThread(threading.Thread):
     def __init__(self, rpc, bug_system, execution, bug_id):
         """
             :param rpc: Bug tracker RPC object
+            :type rpc: dict
             :param bug_system: BugSystem object
+            :type bug_system: dict
```
Comment (atodorov) [inline tcms/issuetracker/base.py:20]: tcms.testcases.models.BugSystem not dict, see Sphinx syntax for making class names links in the docs

File: tcms/issuetracker/base.py [hunk 2]
```diff
@@ -15,9 +15,13 @@ class IntegrationThread(threading.Thread):
     def __init__(self, rpc, bug_system, execution, bug_id):
         """
             :param rpc: Bug tracker RPC object
+            :type rpc: dict
```
Comment (atodorov) [inline tcms/issuetracker/base.py:18]: This is object, not a dict. Actual type is different for each integration class. That's the object facilitating communication with the remote system.

File: tcms/issuetracker/base.py [hunk 3]
```diff
@@ -15,9 +15,13 @@ class IntegrationThread(threading.Thread):
...
             :param bug_system: BugSystem object
+            :type bug_system: dict
             :param execution: TestExecution object
+            :type execution: dict
```
Comment (atodorov) [inline tcms/issuetracker/base.py:22]: TestExecution object, not a dict.

File: tcms/issuetracker/base.py [hunk 4]
```diff
@@ -15,9 +15,13 @@ class IntegrationThread(threading.Thread):
...
             :param bug_id: Unique defect identifier in the system. Usually an int.
+            :type bug_id: int
```
Comment (atodorov) [inline tcms/issuetracker/base.py:24]: usually an int but could be a string too, e.g. JIRA

File: tcms/rpc/api/attachment.py [hunk 5]
```diff
@@ -20,8 +20,8 @@ def remove_attachment(attachment_id, **kwargs):
 
         :param attachment_id: PK of attachment to remove
         :type attachment_id: int
-        :return: None
-        :raises: Exception if attachment doesn't exist,
+        :param kwargs: Dict providing the current HTTP request
```
Comment (atodorov) [inline tcms/rpc/api/attachment.py:23]: Correct but it provides more than the current HTTP request, see https://django-modern-rpc.readthedocs.io/en/latest/configuration_guide/methods_registration.html#access-request-protocol-and-other-info-from-a-rpc-method

Comment (atodorov) [reply tcms/rpc/api/attachment.py:23]: No :type: definition for kwargs ?

File: tcms/rpc/api/component.py [hunk 6]
```diff
@@ -42,6 +42,8 @@ def create(values, **kwargs):
 
         :param values: Field values for :class:`tcms.management.models.Component`
         :type values: dict
+
```
Comment (atodorov) [inline tcms/rpc/api/component.py:45]: Let's go without newline to be consistent with the rest of the doc-strings.

File: tcms/rpc/api/testcase.py [hunk 7]
```diff
@@ -49,6 +49,7 @@ def add_component(case_id, component):
         :param component: Name of Component to add
         :type component: str
         :return: Serialized :class:`tcms.testcases.models.TestCase` object
+        :rtype: dict
         :raises: PermissionDenied if missing the *testcases.add_testcasecomponent*
```
Comment (atodorov) [inline tcms/rpc/api/testcase.py:53]: hmmm, some doc-strings are `:raises Exception: description` while this one and others are `:raises: Exception <description>`. All of them need to be consistent but the linter doesn't seem to prefer one style over the other. 

Is there a difference between the two syntaxes ? If not I prefer to keep the current syntax in order to minimize the changes. If the new syntax is indeed what is needed then it should be changed in other places too but that doesn't seem related to the new linter.

File: tcms/rpc/api/testcase.py [hunk 8]
```diff
@@ -337,6 +336,7 @@ def remove(query):
         :param query: Field lookups for :class:`tcms.testcases.models.TestCase`
         :type query: dict
         :return: None
+        :rtype: NoneType
```
Comment (atodorov) [inline tcms/rpc/api/testcase.py:339]: In many instances above the :return: None reference has been remove but here you specify the return type. Why the difference ? IMO all of these should be consistent.

Comment (brymut) [reply tcms/rpc/api/testcase.py:339]: Apologies, I had switched over later on and missed that I had done this already. I've removed the reference

File: tcms/rpc/api/testcase.py (hunk 7 above)
Comment (brymut) [reply tcms/rpc/api/testcase.py:53]: I remember the linter bringing up errors for some exceptions over this, will comb through again to find the offending ones.

File: tcms/issuetracker/base.py (hunk 2 above)
Comment (brymut) [reply tcms/issuetracker/base.py:18]: So what should I say the type is here then?

File: tcms/rpc/api/testcase.py (hunk 7 above)
Comment (brymut) [reply tcms/rpc/api/testcase.py:53]: I've checked again and it seems like the linter only ever brings up errors if the Exception is explicitly raised in the method itself, unlike here where the `@permissions_required` would be the one to raise it or in other places where the model's manager querying the database would raise the exception. I guess the linter wouldn't be able to infer that there is a PerssionDenied exception being raised here.

File: tcms/rpc/api/attachment.py (hunk 5 above)
Comment (brymut) [reply tcms/rpc/api/attachment.py:23]: Linter doesn't bring up any errors.

File: tcms/issuetracker/base.py (hunk 2 above)
Comment (atodorov) [reply tcms/issuetracker/base.py]: This is of type `object` in the base class and then a specific type coming from a 3rd party library in the actual class providing the implementation.

File: tcms/issuetracker/base.py [hunk 9]
```diff
@@ -15,9 +15,13 @@ class IntegrationThread(threading.Thread):
     def __init__(self, rpc, bug_system, execution, bug_id):
         """
             :param rpc: Bug tracker RPC object
+            :type rpc: :class:`tcms.`
```
Comment (atodorov) [inline tcms/issuetracker/base.py:18]: this is type `object`

Comment (brymut) [reply tcms/issuetracker/base.py]: @atodorov changed.

**Pull request merged**